test/Tests/src/Semantic/Equals_Spec.enso

```diff
@@ -0,0 +1,128 @@
+            _ -> Error.throw ("Unexpected type " + self.to_text)
+
+
+create_random_tree : Integer -> Java_Random -> Integer
+create_random_tree max_depth rnd cur_depth=0 =
+    if cur_depth == max_depth then Node.Nil else
+        next_depth = cur_depth + 1
+        children_count = rnd.nextInt 6
+        children = (0.up_to children_count).map _-> @Tail_Call create_random_tree max_depth rnd next_depth
+        case children_count of
+            0 -> Node.Value 42
+            1 -> Node.C1 (children.at 0)
+            2 -> Node.C2 (children.at 0) (children.at 1)
+            3 -> Node.C3 (children.at 0) (children.at 1) (children.at 2)
+            4 -> Node.C4 (children.at 0) (children.at 1) (children.at 2) (children.at 3)
+            5 -> Node.C5 (children.at 0) (children.at 1) (children.at 2) (children.at 3) (children.at 4)
+            _ -> Error.throw ("Unexpected number of children: " + children_count.to_text)
+
+spec =
+    Test.group "Operator ==" <|
+        Test.specify "should dispatch to overriden `==` on atoms" <|
+            child1 = Child.Value 11
+            parent1 = Parent.Value child1
+            grand_parent1 = GrandParent.Value parent1
+
+            child2 = Child.Value 111
+            parent2 = Parent.Value child2
+            grand_parent2 = GrandParent.Value parent2
+
+            grand_parent1.should_equal grand_parent2
+
+        Test.specify "should handle `==` on types with many fields" <|
+            many_fields1 = ManyFieldType.Value (Child.Value 1) (Child.Value 2) (Child.Value 3) (Child.Value 4) (Child.Value 5) (Child.Value 6) (Child.Value 7) (Child.Value 8) (Child.Value 9) (Child.Value 10) (Child.Value 11) (Child.Value 12) (Child.Value 13) (Child.Value 14) (Child.Value 15)
+            many_fields2 = ManyFieldType.Value (Child.Value 101) (Child.Value 102) (Child.Value 103) (Child.Value 104) (Child.Value 105) (Child.Value 106) (Child.Value 107) (Child.Value 108) (Child.Value 109) (Child.Value 110) (Child.Value 111) (Child.Value 112) (Child.Value 113) (Child.Value 114) (Child.Value 115)
+
+            many_fields1.should_equal many_fields2
+
+        Test.specify "should dispatch to equals on host values" <|
+            java_object1 = HashMap.new
+            java_object1.put "a" 1
+            java_object1.put "b" 2
+
+            java_object2 = HashMap.new
+            java_object2.put "b" 2
+            java_object2.put "a" 1
+
+            java_object1.should_equal java_object2
+
+        Test.specify "should return False for different Atoms with same fields" <|
+            p1 = Point.Value 1 2
+            p2 = Point.Value 3 4
+            rect = Rect.Value p1 p2
+            four_field = FourFieldType.Value 1 2 3 4
+            rect.should_not_equal four_field
+
+        Test.specify "should handle recursive atoms without custom `==`" <|
+            rnd = (Random.new seed=42).java_random
```

Comment:
@radeusgd but `(Random.new seed=42).java_random` initializes the Random java generator with seed `42`, no?

> but please use with_clue to record the seed in case the test fails.

Sure

Reply:
It does, I was sugesting we keep it that way instead of switching to random seed as @JaroslavTulach suggested.

Reply:
It does, I was sugesting we keep it that way instead of switching to random seed as @JaroslavTulach suggested.

Reply:
It does, I was sugesting we keep it that way instead of switching to random seed as @JaroslavTulach suggested.